encoding/string_encode.go

```diff
@@ -1,30 +1,182 @@
 package encoding
 
 import (
-	"errors"
+	"encoding/binary"
+	"fmt"
 	"io"
+	"math"
+
+	"github.com/valyala/gozstd"
+
+	"github.com/xichen2020/eventdb/generated/proto/encodingpb"
+	"github.com/xichen2020/eventdb/x/bytes"
+	"github.com/xichen2020/eventdb/x/unsafe"
+)
+
+// Default initial size of data slice.
+const (
+	defaultInitalDataSliceSize = 1 << 8
+)
+
+// Maximum cardinality allowed for dictionary encoding.
+var (
+	dictEncodingMaxCardinality = 1 << 16
 )
 
-// StringEncoder encodes string values.
+// StringEncoder encodes string valuesIt.
 type StringEncoder interface {
 	// Encode encodes a collection of strings and writes the encoded bytes to the writer.
 	// Callers should explicitly call `Reset` before subsequent call to `Encode`.
-	Encode(writer io.Writer, values ForwardStringIterator) error
-
-	// Reset resets the encoder.
-	Reset()
+	Encode(writer io.Writer, valuesIt RewindableStringIterator) error
 }
 
 // StringEnc is a string encoder.
-type StringEnc struct{}
+type StringEnc struct {
+	buf             []byte
+	metaProto       encodingpb.StringMeta
+	dictionaryProto encodingpb.StringArray
+	data            []string
+}
 
 // NewStringEncoder creates a new string encoder.
-func NewStringEncoder(writer io.Writer) *StringEnc { return &StringEnc{} }
+func NewStringEncoder() *StringEnc {
+	return &StringEnc{
+		buf: make([]byte, 8),
+		metaProto: encodingpb.StringMeta{
+			// Default compression type.
+			Compression: encodingpb.CompressionType_ZSTD,
+		},
+		data: make([]string, defaultInitalDataSliceSize),
+	}
+}
 
 // Encode encodes a collection of strings and writes the encoded bytes to the writer.
-func (enc *StringEnc) Encode(writer io.Writer, values ForwardStringIterator) error {
-	return errors.New("not implemented")
+func (enc *StringEnc) Encode(
+	writer io.Writer,
+	valuesIt RewindableStringIterator,
+) error {
+	// TODO(bodu): Do some perf benchmarking to see whether we want to allocate a new map
+	// or clear an existing one.
+	dictionary := make(map[string]int64)
```

Comment:
Do we want to allocate `dictEncodingMaxCardinality` upfront so we don't reallocate later?

Reply:
👍